packages/compass-editor/src/json-editor.tsx

```diff
@@ -838,11 +900,45 @@ const BaseEditor = React.forwardRef<EditorRef, EditorProps>(function BaseEditor(
         readOnlyExtension,
         EditorView.updateListener.of((update) => {
           updateEditorContentHeight();
+          const editorText = editor.state.sliceDoc() ?? '';
+
+          if (update.focusChanged && editorViewRef.current?.hasFocus) {
+            const editorView = editorViewRef.current;
+
+            onFocusRef.current?.({ editorView, setContent }, editorText);
+          }
+
           if (update.docChanged) {
-            const text = editor.state.sliceDoc() ?? '';
-            onChangeTextRef.current?.(text, update);
+            onChangeTextRef.current?.(editorText, update);
           }
         }),
+        EditorView.domEventHandlers({
+          paste(event, view) {
+            const initialState = view.state.sliceDoc() ?? '';
+            const clipboardContents =
+              event.clipboardData?.getData('text/plain') ?? '';
+            const selection = view.state.selection.main;
+            const isReplace =
+              selection.anchor === 0 && selection.head === initialState.length;
+
+            callWhenEditorIsReady(view, () => {
+              if (isReplace) {
+                onReplaceRef.current?.(
+                  { editorView: view, setContent },
+                  initialState,
+                  view.state.sliceDoc() ?? ''
+                );
+              } else {
+                onPasteRef.current?.(
+                  { editorView: view, setContent },
+                  clipboardContents,
+                  initialState,
+                  view.state.sliceDoc() ?? ''
+                );
+              }
+            });
```

Comment:
I don't think this is how you should approach it: this provides too much information in arguments (i mentioned this above too), defers the call with a microtask (meaning actually two updates happen instead of one: when user pastes, the initial, "incorrect" input will be visible before being replaced). The only thing editor needs to expose to parent components is the original paste event and as you were doing before, you can do your checks in the option editor component, prevent default (codemirror respects `defaultPrevented`) and dispatch your own editor state update

Reply:
In a normal input text I would agree, but I did this earlier, before this design. The problem is that CodeMirror changes are not sync, they are async, and might trigger events before the Redux state is updated. This happened to me in the test cases, where sometimes it worked, and sometimes not, because there might be race conditions. Using the state from CodeMirror ensures that there is only one source of truth.

Does that make sense?

Reply:
(Sorry missed this comment previously) While not explicitly called out in the docs, you can look at the codemirror source to confirm that [state updates are synchronous](https://github.com/codemirror/view/blob/464eb9e6de5f430d1cf82c62aad95c20fb70d345/src/editorview.ts#L227-L345), the problem arises when you are starting to dispatch events from the callbacks that are tied to the editor lifecycle, as this happens internally while update is still in progress, you run into this issue that we introduced the `scheduleUpdate` method for. In all other cases the order of dispatch is preserved, but there are some slight variations based on the provided `TransactionSpec` to how they are applied, [see `sequential` for example](https://github.com/codemirror/state/blob/84d75c37a8c546181463ef286e31e48b165b4168/src/transaction.ts#L147-L153), even in this case it's only inside one state update that this property takes effect.

Also, just once again to highlight, I mentioned this in other comments also: redux state is not relevant here, yes, it will be not always in sync with editor state and yes, you don't need to rely on it for this change, but it's also not the reason to handle it inside the editor component like that